### PR TITLE
update GitHub Actions checkout, setup, and codecov versions to v3

### LINKED ISSUE #734: Github Actions Update
Saw this while looking at our job queue.

```
Annotations
1 warning
build
Node.js 12 actions are deprecated. For more information see: https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/. Please update the following actions to use Node.js 16: actions/checkout, actions/checkout
```

### LINKED ISSUE #752: Another GitHub Actions Deprecation warning
https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

### PR DESCRIPTION
**Pull request by garrettwrong**
Updates for github actions version ~and rolls stable Python from 3.7 to 3.8, and 3.11~ rolls conda env to 3.8.

Closes #734

### CODE REVIEW

**Comment by garrettwrong** [comment]
Probably Closes #752

**Comment by garrettwrong** [comment]
Looks like we'll need FINUFFT for 3.11.

https://github.com/flatironinstitute/finufft/issues/245

**Comment by garrettwrong** [comment]
I was too aggressive with this PR.  Python 3.11 is not ready on all Anaconda platforms or ready for `Ray` yet.

I am going to rebase this PR, saving the 3.11 changes in a different branch for later.

**Comment by garrettwrong** [comment]
Note that this will require updating the GitHub "Status Checks" and updating all outstanding branches once merged.

File: README.md
Comment (chris-langfield) [inline README.md]: Should we say requires Python 3.8 if we're encouraging using the `conda` env?

Comment (chris-langfield) [reply README.md]: I think just the two 3.7's in these two sentences should change the 3.8, but the rest are okay because we are still supporting, and they can use pip with >= 3.7 etc

**Review by chris-langfield — CHANGES_REQUESTED** [review]

File: README.md
Comment (garrettwrong) [reply README.md]: ASPIRE requires >=3.7 for pip.  You are right that this is somewhat misleading. I will change it (similar to installation.rst) Let me know if it works better.

Comment (garrettwrong) [reply README.md]: (We want it to be accurate, but also keep in mind if we invert the conda/pip version pinning relationship that the docs will need to be updated again anyway).

File: docs/source/installation.rst
Comment (j-c-c) [inline docs/source/installation.rst]: Being picky, but I think "reproducing" sounds better than "reconstructing" and also has a better connotation. Not super important either way.

**Review by j-c-c — DISMISSED** [review]
LGTM! Left one comment that's up to you if you want to change.

File: docs/source/installation.rst
Comment (garrettwrong) [reply docs/source/installation.rst]: Sure, will change.

**Review by chris-langfield — APPROVED** [review]
Definitely clearer thanks!

**Review by j-c-c — APPROVED** [review]

**Comment by garrettwrong** [comment]
Going to merge this and adjust the CI checks once the Azure thing settles.

**Pull request merged**
